### ISSUE TITLE
Apollo polling does not continue after socket closed  graceful reconnect

### ISSUE THREAD

**Issue by D1plo1d** [enisdenjo/graphql-ws#137]
Simplifying some details my code looks aproximately something like this:
```
  const { data, error, previousData } = useQuery(LATENCY_QUERY, {
    pollInterval: PING_INTERVAL,
  })
  useEffect(() => console.log(data), [data])
```

After I shut down my server and restart it (causing the socket to close and gracefully re-open) I stop seeing polling data in my console. How do I get apollo client to restart polling on re-connect?

**Comment by D1plo1d** [comment]
Edit: Clicked submit too soon - details in the issue summary now

**Comment by enisdenjo** [comment]
Hey there! I'll start off with two questions: Does the WebSocket successfully reconnect? You sure `useQuery` causes a re-render with new data for the effect hook?

Sadly, I've never used Apollo myself, so I really don't know the inner workings of the `useQuery`. What makes you think this is an issue with `graphql-ws` over Apollo's client?

**Comment by D1plo1d** [comment]
Hello! Yes, the connection is re-established as evidenced by my being able to make other queries and mutations. The query here returns the current server timestamp in millis and is polled once every 500ms so every response contains data that should (and before reconnection does) case `useQuery` to re-render.

So regarding whether this is an apollo bug or a graphql-ws bug I'm not as clear. It's certainly an apollo integrated with graphql-ws bug but I am not sure beyond that!

**Comment by enisdenjo** [comment]
Hmm, what about the WebSocketLink? Does the `request` method get triggered?

If you're sure about this being a graphql-ws + apollo bug, I would be very appreciative if you could provide me with a minimal repro -  I'd need to fiddle around in order to figure this one out because I have zero experience with the Apollo stack.

**Comment by D1plo1d** [comment]
Ah, that's a good question. Let me throw some console.logs in the link and try debugging this from that angle. If I still come up with nothing I'll shoot you a minimal repo later this week.

**Comment by enisdenjo** [comment]
If the relevant link gets invoked, the WS client `subscription` method gets called, but none of the callbacks get triggered; then its likely that `graphql-ws` is swallowing requests, which is bad! I'll jump onto this ASAP.

However, if the link is never invoked - I'd guess this issue lies on Apollo's side.

**Comment by D1plo1d** [comment]
Yeah, makes sense. Appreciate it!

**Comment by D1plo1d** [comment]
Little update, the requests did stop after the reconnect so it is looking like it is an apollo issue.. that and there's an issue about pretty much  this exact thing I failed to see last time :sweat_smile:  https://github.com/apollographql/apollo-client/issues/1099

Closing this issue. Thanks again!

**Comment by enisdenjo** [comment]
Might be, however, I would like to know one more thing: does the poll query coming before the silent reconnect settle? Could there be a stuck query which prevents Apollo from issuing a new one?

**Comment by D1plo1d** [comment]
Yeah so I did more digging and found some interesting results regarding that but I'm using a (*cough* very) custom socket implementation which could be to blame. This is my "websocket": https://github.com/tegapp/teg/blob/develop/packages/teg-web-ui/src/webrtc/WebRTCSocket.ts

That caveat out of the way I found that I received a `completed` event for my polling queries on disconnect but no `next` or `error` events. I think this lines up with your stuck query theory, yeah?

**Comment by D1plo1d** [comment]
I found the best workaround was to add a callback to my connection code. When a connection succeeds a counter is incremented via `setReconnectCounter(i => i + 1)` causing a re-rerender where I use the counter value as the key for my apollo provider: `<ApolloProvider client={client} key={reconnectCounter}>`. After that the entire React component tree gets remounted so any weirdness in stuck queries gets wiped out in the remount.

**Comment by enisdenjo** [comment]
As per the Observer pattern, calling a `complete` or `error` ends the observable. So, having called the `complete` on the subscription indicates that the request is done and a new one should be issued afterwards. If this is a case with the last poll before disconnect, it is indeed an Apollo bug.

**Comment by enisdenjo** [comment]
Or maybe the WebRTC implementation is at play. To debunk that, we'd need to know if the accompanying messages are being issued through the socket and if the implementation is calling the right callbacks.

If you could supply me with a repro, I could check that out too. 😄

**Comment by enisdenjo** [comment]
@D1plo1d any news?

**Comment by D1plo1d** [comment]
Hey sorry, once I got the workaround working I got swamped with work getting a release out the door. I'm happy to say it was probably the WebRTC implementation for now and leave it there unless someone else notices anything similar with a Websocket.

My repo is [here but it's 25,000+ lines of code!](https://github.com/tegapp/teg) I wouldn't wish it upon you to have to sort through all that!

Thanks again for your help debugging this! Very appreciated!
